### PR TITLE
[minor] Fixes #115

### LINKED ISSUE #115: [mig client] Add option for target agents from results of previous action (Bugzilla #1141910)
Migrated from https://bugzilla.mozilla.org/show_bug.cgi?id=1141910
Assigned to: Julien Vehent [:ulfr]

On 2015-03-10 20:56:03 -0700, Julien Vehent [:ulfr] wrote:

> Add options that automatically implements http://mig.mozilla.org/doc/cheatsheet.rst.html#target-agents-that-found-results-in-a-previous-action
> 
> mig file -t "id IN ( \
>     SELECT agentid FROM commands, json_array_elements(commands.results) AS r \
>     WHERE commands.actionid = 12345 AND r#>>'{foundanything}' = 'true')" \
> -path /etc/passwd -content "^spongebob"
> 
> For example:
> * --target-foundanything <actionid> would limit the targets to agents that returned positive results in <actionid>
> * --target-foundnothing <actionid> would do the opposite and target agents that did not return positive results in <actionid>
> 
> This option should concatenate with the regular -t option such that an investigator could set both and filter the results even further.

On 2015-03-10 20:56:41 -0700, Julien Vehent [:ulfr] wrote:

> Implementation should take the form of two new options in the command line client at https://github.com/mozilla/mig/blob/master/src/mig/client/cmd/main.go#L75

### PR DESCRIPTION
**Pull request by arunk-s**
Add `-target-foundanything` and `-targetfoundnothing` flags to mig client

### CODE REVIEW

File: client/mig/main.go [hunk 1]
```diff
@@ -223,6 +230,20 @@ func main() {
 	for _, arg := range os.Args[1:] {
 		a.Name += arg + " "
 	}
+
+	if targetfound != "" && targetnotfound != "" {
+		panic("Both -target-foundanything and -target-foundnothing cannot be used simultaneously")
+	}
+	if targetfound != "" {
+		targetQuery := fmt.Sprintf(`id IN (select agentid from commands, json_array_elements(commands.results) as 
+			r where actionid=%s and r#>>'{foundanything}' = 'true')`, targetfound)
```
Comment (jvehent) [inline client/mig/main.go]: we probably don't want a newline in here, as the string will be added to the json

File: client/mig/main.go [hunk 2]
```diff
@@ -223,6 +230,20 @@ func main() {
...
+		target = targetQuery + " AND " + target
+	}
+	if targetnotfound != "" {
+		targetQuery := fmt.Sprintf(`id NOT IN (select agentid from commands, json_array_elements(commands.results) as 
+			r where actionid=%s and r#>>'{foundanything}' = 'true')`, targetnotfound)
```
Comment (jvehent) [inline client/mig/main.go]: same comment about newline

File: client/mig/main.go [hunk 3]
```diff
@@ -94,6 +99,8 @@ func main() {
 	fs.StringVar(&show, "show", "found", "type of results to show")
 	fs.StringVar(&render, "render", "text", "results rendering mode")
 	fs.StringVar(&target, "t", fmt.Sprintf("status='%s' AND mode='daemon'", mig.AgtStatusOnline), "action target")
+	fs.StringVar(&targetfound, "target-foundanything", "", "Limit targets to agents that returned positive results")
+	fs.StringVar(&targetnotfound, "target-foundnothing", "", "Limit targets to agents that did not returned positive results")
```
Comment (jvehent) [inline client/mig/main.go]: These explanations are vague. What about `targets agents that have found results in a previous action. ex: -target-found 123456`.

File: client/mig/main.go [hunk 4]
```diff
@@ -94,6 +99,8 @@ func main() {
 	fs.StringVar(&show, "show", "found", "type of results to show")
 	fs.StringVar(&render, "render", "text", "results rendering mode")
 	fs.StringVar(&target, "t", fmt.Sprintf("status='%s' AND mode='daemon'", mig.AgtStatusOnline), "action target")
+	fs.StringVar(&targetfound, "target-foundanything", "", "Limit targets to agents that returned positive results")
```
Comment (jvehent) [inline client/mig/main.go]: we can remove the `anything` part of the string, `target-found` is enough.

File: client/mig/main.go (hunk 1 above)
Comment (arunk-s) [inline client/mig/main.go]: Oh, I thought that by line wrapping you meant that I should divide the line into two.
Should I rather be doing this:
```go
targetQuery := fmt.Sprintf(`id IN (select agentid from commands, json_array_elements(commands.results) as ` +
`r where actionid=%s and r#>>'{foundanything}' = 'true')`, targetfound) 
```
which divide the line into two and also doesn't add a newline.

File: client/mig/main.go (hunk 3 above)
Comment (arunk-s) [inline client/mig/main.go]: Yeah. That makes more sense. I'll update the code to reflect that.

**Comment by jvehent** [comment]
r+, thanks for the patch :)

**Comment by jvehent** [comment]
I'm fixing bugs with this code in another commit.

**Pull request merged**